[LFC][IFC] Remove floats that are outside of the "rebuild" range to ensure we don't add them twice

obyknovenius opened the pull request:
#### 363d7e6b674dc92a23cf41e34737254b0ac81707
<pre>
[LFC][IFC] Remove floats that are outside of the &quot;rebuild&quot; range to ensure we don&apos;t add them twice
<a href="https://bugs.webkit.org/show_bug.cgi?id=247625">https://bugs.webkit.org/show_bug.cgi?id=247625</a>

Reviewed by Alan Baradlay.

When rebuilding a line, we should remove any floats starting from
`lastInlineItemToAdd` up to `layoutRange.endIndex()` from both
`layoutState().placedFloats()` and `m_placedFloats`.

* LayoutTests/TestExpectations:
* LayoutTests/fast/block/float/intrusive-floats-with-reverted-inline-content-expected.txt:
* LayoutTests/fast/block/float/intrusive-floats-with-reverted-inline-content.html:
* Source/WebCore/layout/floats/PlacedFloats.h:
(WebCore::Layout::PlacedFloats::remove):
* Source/WebCore/layout/formattingContexts/inline/InlineLineBuilder.cpp:
(WebCore::Layout::LineBuilder::rebuildLineWithInlineContent):

Canonical link: <a href="https://commits.webkit.org/282129@main">https://commits.webkit.org/282129@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1ec73acc8ef30c041fbcb610adf98f8582d05c45

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62151 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41505 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14743 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66131 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12696 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64270 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49192 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13036 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50089 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8793 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65220 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38575 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53830 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30892 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35226 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11095 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11627 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56962 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11399 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67861 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6094 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11169 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57467 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6120 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53805 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57691 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13817 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5032 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37305 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38389 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39485 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38134 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineLineBuilder.cpp:
```diff
@@ -1186,6 +1198,12 @@ size_t LineBuilder::rebuildLineWithInlineContent(const InlineItemRange& layoutRa
+    // Remove floats that are outside of this "rebuild" range to ensure we don't add them twice.
+    for (; index < layoutRange.endIndex(); ++index) {
+        if (m_inlineItemList[index].isFloat())
+            unplaceFloatBox(m_inlineItemList[index].layoutBox());
```
undoing float placement is very specific to this "rebuild line" use case so I think it'd better if this functionality stayed private to this function. 
I think it's fine to have a simple `PlacedFloats::remove(const Box&) `and just call `removeFirst`.
-also this loop can bail out early on the first float that has not been placed yet.

obyknovenius replied:
All comments have been fixed.

obyknovenius committed:
[LFC][IFC] Remove floats that are outside of the "rebuild" range to ensure we don't add them twice
https://bugs.webkit.org/show_bug.cgi?id=247625

Reviewed by Alan Baradlay.

When rebuilding a line, we should remove any floats starting from
`lastInlineItemToAdd` up to `layoutRange.endIndex()` from both
`layoutState().placedFloats()` and `m_placedFloats`.

* LayoutTests/TestExpectations:
* LayoutTests/fast/block/float/intrusive-floats-with-reverted-inline-content-expected.txt:
* LayoutTests/fast/block/float/intrusive-floats-with-reverted-inline-content.html:
* Source/WebCore/layout/floats/PlacedFloats.h:
(WebCore::Layout::PlacedFloats::remove):
* Source/WebCore/layout/formattingContexts/inline/InlineLineBuilder.cpp:
(WebCore::Layout::LineBuilder::rebuildLineWithInlineContent):

Canonical link: https://commits.webkit.org/282129@main
